libraries/splunk_app.rb

```diff
@@ -126,13 +174,80 @@ def create_splunk_directory(path)
         dir.group(node['splunk']['group'])
         dir.mode('0755')
         dir.run_action(:create)
-        new_resource.updated_by_last_action(dir.updated_by_last_action?)
+        new_resource.updated_by_last_action(true) if dir.updated_by_last_action?
+      end
+
+      def download_and_install # rubocop:disable PerceivedComplexity, CyclomaticComplexity
+        return if new_resource.url.nil? || new_resource.url.empty?
+
+        expected_version = CernerSplunk::AppVersion.new new_resource.version
+        installed_version = CernerSplunk::AppVersion.new @current_resource.version
+
+        return if expected_version.type && expected_version == installed_version
+        Chef::Log.warn "Expected version not specified for #{new_resource.app}." unless expected_version.type
+        fail "Expecting to install prerelease on top of same released version for #{new_resource.app}" if expected_version.type == :prerelease && installed_version.type == :base && expected_version.base == installed_version.base
+
+        # In whyrun mode, the rest is not safe.
+        return if whyrun_mode?
+
+        filename = "#{Chef::Config[:file_cache_path]}/#{new_resource.app}.tgz"
+
+        download = Chef::Resource::RemoteFile.new(filename, run_context)
+        download.source(new_resource.url)
+        download.run_action(:create)
+        download.backup(false)
+
+        install_from_tar filename, expected_version, installed_version
+      ensure
+        download.run_action(:delete) if download
+      end
+
+      def install_from_tar(filename, expected_version, installed_version) # rubocop:disable PerceivedComplexity, CyclomaticComplexity
```

Comment:
This is doing a lot.  Is there a way to split out some of the reading into helper functions? 

Like move the untaring to it's own function and the validation into it's own function? Just thoughts to maybe make this a little easier for humans to parse.  At least this human,  :smiley_cat: .